### PR TITLE
[Mailer] Move abstract test case to Tests/

### PR DESCRIPTION
**Pull request by OskarStark**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | ---
| License       | MIT
| Doc PR        | ---

This PR

is based on https://github.com/symfony/symfony/pull/39495/files#r544203249

cc @wouterj @nicolas-grekas

### CODE REVIEW

File: .github/composer-config.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
         "preferred-install": {
             "symfony/form": "source",
             "symfony/http-kernel": "source",
+            "symfony/mailer": "source",
```
Comment (OskarStark) [inline .github/composer-config.json:7]: Is this correct and needed?

File: src/Symfony/Component/Mailer/Bridge/Amazon/composer.json [hunk 2]
```diff
@@ -20,7 +20,7 @@
         "symfony/mailer": "^4.4|^5.0"
     },
     "require-dev": {
-        "symfony/http-client": "^4.3|^5.0"
+        "symfony/mailer": "^4.4.18|^5.1.10"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/Bridge/Amazon/composer.json]: httpclient is not needed here?
note that mailer is already in require - it should not be duplicated inrequire-dev

File: src/Symfony/Bundle/FrameworkBundle/Test/MailerAssertionsTrait.php [hunk 3]
```diff
@@ -14,7 +14,7 @@
 use PHPUnit\Framework\Constraint\LogicalNot;
 use Symfony\Component\Mailer\Event\MessageEvent;
 use Symfony\Component\Mailer\Event\MessageEvents;
-use Symfony\Component\Mailer\Test\Constraint as MailerConstraint;
+use Symfony\Component\Mailer\Tests\Constraint as MailerConstraint;
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Test/MailerAssertionsTrait.php]: this one should not be changed

File: .github/composer-config.json (hunk 1 above)
Comment (nicolas-grekas) [reply .github/composer-config.json:7]: it is needed yes, to install the Tests/ dir in vendors when testing transports

File: src/Symfony/Component/Mailer/Bridge/Google/composer.json [hunk 4]
```diff
@@ -20,7 +20,7 @@
         "symfony/mailer": "^4.4|^5.0"
     },
     "require-dev": {
-        "symfony/http-client": "^4.3|^5.0"
+        "symfony/mailer": "^4.4.18|^5.1.10"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/Bridge/Google/composer.json]: same question

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/composer.json [hunk 5]
```diff
@@ -20,7 +20,7 @@
         "symfony/mailer": "^4.4|^5.0"
     },
     "require-dev": {
-        "symfony/http-client": "^4.3|^5.0"
+        "symfony/http-client": "^4.4.18|^5.1.10"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/Bridge/Mailchimp/composer.json]: why bump?

File: src/Symfony/Component/Mailer/Bridge/Mailgun/composer.json [hunk 6]
```diff
@@ -20,7 +20,7 @@
         "symfony/mailer": "^4.4|^5.0"
     },
     "require-dev": {
-        "symfony/http-client": "^4.3|^5.0"
+        "symfony/mailer": "^4.4.18|^5.1.10"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/Bridge/Mailgun/composer.json]: same question

File: src/Symfony/Component/Mailer/Bridge/Postmark/composer.json [hunk 7]
```diff
@@ -20,7 +20,7 @@
         "symfony/mailer": "^4.4|^5.0"
     },
     "require-dev": {
-        "symfony/http-client": "^4.3|^5.0"
+        "symfony/mailer": "^4.4.18|^5.1.10"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/Bridge/Postmark/composer.json]: same question

File: src/Symfony/Component/Mailer/Bridge/Sendgrid/composer.json [hunk 8]
```diff
@@ -20,7 +20,7 @@
         "symfony/mailer": "^4.4|^5.0"
     },
     "require-dev": {
-        "symfony/http-client": "^4.3|^5.0"
+        "symfony/mailer": "^4.4.18|^5.1.10"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/Bridge/Sendgrid/composer.json]: same question

File: src/Symfony/Component/Mailer/Test/Constraint/EmailCount.php [hunk 9]
```diff
@@ -9,7 +9,7 @@
  * file that was distributed with this source code.
  */
 
-namespace Symfony\Component\Mailer\Test\Constraint;
+namespace Symfony\Component\Mailer\Tests\Constraint;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/Test/Constraint/EmailCount.php]: to be reverted

File: src/Symfony/Component/Mailer/Test/Constraint/EmailIsQueued.php [hunk 10]
```diff
@@ -9,7 +9,7 @@
  * file that was distributed with this source code.
  */
 
-namespace Symfony\Component\Mailer\Test\Constraint;
+namespace Symfony\Component\Mailer\Tests\Constraint;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/Test/Constraint/EmailIsQueued.php]: to be reverted

File: src/Symfony/Bundle/FrameworkBundle/Test/MailerAssertionsTrait.php (hunk 3 above)
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/Test/MailerAssertionsTrait.php]:
```suggestion
use Symfony\Component\Mailer\Test\Constraint as MailerConstraint;
```

File: src/Symfony/Component/Mailer/Bridge/Amazon/composer.json (hunk 2 above)
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Amazon/composer.json]:
```suggestion
        "symfony/http-client": "^4.3|^5.0"
```

File: src/Symfony/Component/Mailer/Bridge/Amazon/composer.json [hunk 11]
```diff
@@ -20,7 +20,7 @@
         "symfony/mailer": "^4.4|^5.0"
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Amazon/composer.json]:
```suggestion
        "symfony/mailer": "^4.4.18|^5.1.10"
```

File: src/Symfony/Component/Mailer/Bridge/Google/composer.json [hunk 12]
```diff
@@ -20,7 +20,7 @@
         "symfony/mailer": "^4.4|^5.0"
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Google/composer.json]:
```suggestion
        "symfony/mailer": "^4.4.18|^5.1.10"
```

File: src/Symfony/Component/Mailer/Bridge/Google/composer.json (hunk 4 above)
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Google/composer.json]:
```suggestion
        "symfony/http-client": "^4.3|^5.0"
```

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/composer.json [hunk 13]
```diff
@@ -20,7 +20,7 @@
         "symfony/mailer": "^4.4|^5.0"
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Mailchimp/composer.json]:
```suggestion
        "symfony/mailer": "^4.4.18|^5.1.10"
```

File: src/Symfony/Component/Mailer/Bridge/Mailchimp/composer.json (hunk 5 above)
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Mailchimp/composer.json]:
```suggestion
        "symfony/http-client": "^4.3|^5.0"
```

File: src/Symfony/Component/Mailer/Bridge/Mailgun/composer.json [hunk 14]
```diff
@@ -20,7 +20,7 @@
         "symfony/mailer": "^4.4|^5.0"
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Mailgun/composer.json]:
```suggestion
        "symfony/mailer": "^4.4.18|^5.1.10"
```

File: src/Symfony/Component/Mailer/Bridge/Mailgun/composer.json (hunk 6 above)
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Mailgun/composer.json]:
```suggestion
        "symfony/http-client": "^4.3|^5.0"
```

File: src/Symfony/Component/Mailer/Bridge/Postmark/composer.json [hunk 15]
```diff
@@ -20,7 +20,7 @@
         "symfony/mailer": "^4.4|^5.0"
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Postmark/composer.json]:
```suggestion
        "symfony/mailer": "^4.4.18|^5.1.10"
```

File: src/Symfony/Component/Mailer/Bridge/Postmark/composer.json (hunk 7 above)
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Postmark/composer.json]:
```suggestion
        "symfony/http-client": "^4.3|^5.0"
```

File: src/Symfony/Component/Mailer/Bridge/Sendgrid/composer.json [hunk 16]
```diff
@@ -20,7 +20,7 @@
         "symfony/mailer": "^4.4|^5.0"
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Sendgrid/composer.json]:
```suggestion
        "symfony/mailer": "^4.4.18|^5.1.10"
```

File: src/Symfony/Component/Mailer/Bridge/Sendgrid/composer.json (hunk 8 above)
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Sendgrid/composer.json]:
```suggestion
        "symfony/http-client": "^4.3|^5.0"```

File: src/Symfony/Component/Mailer/Test/Constraint/EmailCount.php (hunk 9 above)
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Test/Constraint/EmailCount.php]:
```suggestion
namespace Symfony\Component\Mailer\Test\Constraint;
```

File: src/Symfony/Component/Mailer/Test/Constraint/EmailIsQueued.php (hunk 10 above)
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Test/Constraint/EmailIsQueued.php]:
```suggestion
namespace Symfony\Component\Mailer\Test\Constraint;
```

File: src/Symfony/Component/Mailer/Bridge/Sendgrid/composer.json [hunk 17]
```diff
@@ -17,10 +17,10 @@
     ],
     "require": {
         "php": ">=7.1.3",
-        "symfony/mailer": "^4.4|^5.0"
+        "symfony/mailer": "^4.4.18|^5.1.10"
     },
     "require-dev": {
-        "symfony/http-client": "^4.3|^5.0"
+        "symfony/http-client": "^4.3|^5.0"```
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Bridge/Sendgrid/composer.json]:
```suggestion
        "symfony/http-client": "^4.3|^5.0"
```

File: src/Symfony/Component/Mailer/Test/Constraint/EmailCount.php (hunk 9 above)
Comment (OskarStark) [inline src/Symfony/Component/Mailer/Test/Constraint/EmailCount.php]:
```suggestion
namespace Symfony\Component\Mailer\Test\Constraint;
```

File: src/Symfony/Component/Mailer/Bridge/Amazon/composer.json (hunk 2 above)
Comment (OskarStark) [reply src/Symfony/Component/Mailer/Bridge/Amazon/composer.json]: Sorry I wanted to bump mailer, not http-client, I mixed sth up

Is it correct to bump mailer like that now?

Comment (nicolas-grekas) [reply src/Symfony/Component/Mailer/Bridge/Amazon/composer.json]: nope: the line in "require" should be updated instead

**Comment by OskarStark** [comment]
@nicolas-grekas I bumped all in `require` section

File: src/Symfony/Component/Mailer/Bridge/Amazon/composer.json [hunk 18]
```diff
@@ -17,7 +17,7 @@
     ],
     "require": {
         "php": ">=7.1.3",
-        "symfony/mailer": "^4.4|^5.0"
+        "symfony/mailer": "^4.4.18|^5.1.10"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/Bridge/Amazon/composer.json]: `"^4.4.18|~5.1.10|^5.2.1"` for all of them

Comment (OskarStark) [reply src/Symfony/Component/Mailer/Bridge/Amazon/composer.json]: Ah damn you are right :-) 

Done

**Comment by xabbuh** [comment]
That's a BC break, isn't it?

**Review by nicolas-grekas — APPROVED** [review]
> That's a BC break, isn't it?

in theory yes, in practice using this `Test` namespace was just a workaround for something that is now fixed by the new line in `.github/composer-config.json`

File: src/Symfony/Component/Mailer/Bridge/Amazon/composer.json [hunk 19]
```diff
@@ -17,7 +17,7 @@
     ],
     "require": {
         "php": ">=7.1.3",
-        "symfony/mailer": "^4.4|^5.0"
+        "symfony/mailer": "^4.4.18|^5.1.10|^5.2.1"
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/Bridge/Amazon/composer.json]: `~5.1.10` instead of `^5.1.10` everywhere

Comment (OskarStark) [reply src/Symfony/Component/Mailer/Bridge/Amazon/composer.json]: Done

**Comment by wouterj** [comment]
I'm :-1: here. We have quite some components offering these abstract test classes to be used outside the Symfony framework. They make testing custom mailer transports easier for everyone. This even was the reason this class was moved in `Test` (from `Tests`) in https://github.com/symfony/symfony/pull/32749

In any case, if this is still approved it *is* a BC break and can't be done in 4.4 imho. But I believe we should keep this as-is and consider the abstract Notifier test case (related PR) a feature.

**Comment by nicolas-grekas** [comment]
Thanks for the context @wouterj this makes perfect sense!
I'm OK then for doing the same in Notifier!

**Comment by wouterj** [comment]
_(I'm sorry if my comment was a bit blunt. With one approval in already, I was in a bit of a rush to prevent quick merge)_

**Comment by OskarStark** [comment]
> (I'm sorry if my comment was a bit blunt. With one approval in already, I was in a bit of a rush to prevent quick merge)

All fine , thanks 👍

**Pull request closed without merging**